File: testing/stub_llm_client.ts
// testing/stub_llm_client.ts — Deterministic ILLMClient for unit tests
import {
  CompletionChunk,
  CompletionRequest,
  CompletionResponse,
  ILLMClient,
  LLMVendor,
  TokenLimitError,
} from '../core/protocols.js'

interface ScriptEntry {
  pattern: RegExp | string
  response: Partial<CompletionResponse>
}

// Scripted responses keyed by message content pattern — §17 CLAUDE.md
export class StubLLMClient implements ILLMClient {
  readonly vendor: LLMVendor
  readonly modelId: string
  readonly maxTokens: number
  readonly supportsTools: boolean

  private readonly _scripts: ScriptEntry[] = []
  private _callCount = 0

  constructor(options: {
    vendor?: LLMVendor
    model?: string
    maxTokens?: number
    supportsTools?: boolean
  } = {}) {
    this.vendor = options.vendor ?? 'anthropic'
    this.modelId = options.model ?? 'stub-1.0'
    this.maxTokens = options.maxTokens ?? 4096
    this.supportsTools = options.supportsTools ?? false
  }

  // Fluent API for scripting responses
  addScript(pattern: RegExp | string, response: Partial<CompletionResponse>): this {
    this._scripts.push({ pattern, response })
    return this
  }

  get callCount(): number { return this._callCount }

  async complete(request: CompletionRequest): Promise<CompletionResponse> {
    this._callCount++
    const estimated = Math.ceil(
      request.messages.reduce((s, m) => s + m.content.length, 0) / 4,
    )
    if (estimated > this.maxTokens) {
      throw new TokenLimitError(estimated, this.maxTokens)
    }

    const lastUser = [...request.messages]
      .reverse()
      .find(m => m.role === 'user')?.content ?? ''

    for (const entry of this._scripts) {
      const matches =
        typeof entry.pattern === 'string'
          ? lastUser.includes(entry.pattern)
          : entry.pattern.test(lastUser)
      if (matches) {
        return {
          content: entry.response.content ?? 'Stub response.',
          stopReason: entry.response.stopReason ?? 'end_turn',
          usage: entry.response.usage ?? { inputTokens: 10, outputTokens: 20 },
          toolCalls: entry.response.toolCalls,
        }
      }
    }

    return {
      content: 'Default stub response.',
      stopReason: 'end_turn',
      usage: { inputTokens: 10, outputTokens: 20 },
    }
  }

  async *stream(request: CompletionRequest): AsyncIterableIterator<CompletionChunk> {
    const response = await this.complete(request)
    yield { delta: response.content, done: false }
    yield { delta: '', done: true }
  }
}

File: tests/stub_llm_client.py
"""tests/stub_llm_client.py — Deterministic ILLMClient for unit tests."""

from __future__ import annotations

import math
from collections.abc import AsyncIterator
from re import Pattern
from typing import Any

from ooagent.core.protocols import (
    CompletionChunk,
    CompletionRequest,
    CompletionResponse,
    ILLMClient,
    LLMVendor,
    TokenLimitError,
    TokenUsage,
)


class _ScriptEntry:
    """Scripted response entry keyed by a message-content pattern."""

    __slots__ = ("pattern", "response")

    def __init__(self, pattern: str | Pattern[str], response: dict[str, Any]) -> None:
        self.pattern = pattern
        self.response = response


class StubLLMClient(ILLMClient):
    """Scripted responses keyed by message content pattern — §17 CLAUDE.md."""

    def __init__(
        self,
        vendor: LLMVendor = "anthropic",
        model: str = "stub-1.0",
        max_tokens: int = 4096,
        supports_tools: bool = False,
    ) -> None:
        self._vendor = vendor
        self._model_id = model
        self._max_tokens = max_tokens
        self._supports_tools = supports_tools
        self._scripts: list[_ScriptEntry] = []
        self._call_count = 0

    def add_script(self, pattern: str | Pattern[str], response: dict[str, Any]) -> StubLLMClient:
        """Fluent API for scripting responses."""
        self._scripts.append(_ScriptEntry(pattern, response))
        return self

    @property
    def call_count(self) -> int:
        return self._call_count

    @property
    def model_id(self) -> str:
        return self._model_id

    @property
    def vendor(self) -> LLMVendor:
        return self._vendor

    @property
    def max_tokens(self) -> int:
        return self._max_tokens

    @property
    def supports_tools(self) -> bool:
        return self._supports_tools

    async def complete(self, request: CompletionRequest) -> CompletionResponse:
        self._call_count += 1
        estimated = math.ceil(sum(len(m.content) for m in request.messages) / 4)
        if estimated > self._max_tokens:
            raise TokenLimitError(estimated, self._max_tokens)

        last_user = ""
        for message in reversed(request.messages):
            if message.role == "user":
                last_user = message.content
                break

        for entry in self._scripts:
            if isinstance(entry.pattern, str):
                matches = entry.pattern in last_user
            else:
                matches = entry.pattern.search(last_user) is not None
            if matches:
                usage = entry.response.get("usage") or {
                    "input_tokens": 10,
                    "output_tokens": 20,
                }
                return CompletionResponse(
                    content=entry.response.get("content", "Stub response."),
                    stop_reason=entry.response.get("stop_reason", "end_turn"),
                    usage=TokenUsage(**usage) if isinstance(usage, dict) else usage,
                    tool_calls=entry.response.get("tool_calls"),
                )

        return CompletionResponse(
            content="Default stub response.",
            stop_reason="end_turn",
            usage=TokenUsage(input_tokens=10, output_tokens=20),
        )

    async def stream(self, request: CompletionRequest) -> AsyncIterator[CompletionChunk]:
        response = await self.complete(request)
        yield CompletionChunk(delta=response.content, done=False)
        yield CompletionChunk(delta="", done=True)
